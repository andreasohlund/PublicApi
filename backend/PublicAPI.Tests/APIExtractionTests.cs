using ApprovalTests.Reporters;

[assembly: UseReporter(typeof(DiffReporter))]

namespace PublicAPI.Tests
{
    using ApprovalTests;
    using NUnit.Framework;
    using PublicAPI.APIExtraction;
    using System;
    using System.IO;
    using System.Net.Http;
    using System.Text.Json;
    using System.Threading.Tasks;

    [TestFixture]
    public class APIExtractionTests
    {
        [Test]
        public async Task ApproveAPIExtraction()
        {
            var extractor = new AssemblyAPIExtractor();
            var binFolder = AppDomain.CurrentDomain.BaseDirectory.Replace("PublicAPI.Tests", "ExampleAssembly");
            var path = Path.Combine(binFolder, "ExampleAssembly.dll");

            using var assemblyStream = new FileStream(path, FileMode.Open);

            var publicTypes = await extractor.ExtractFromStream(assemblyStream);

            Approvals.VerifyJson(JsonSerializer.Serialize(publicTypes));
        }

        [Test]
        public async Task ApproveNServiceBus710API()
        {
            var packageId = "NServiceBus";
            var version = "7.1.0";

            var httpClient = new HttpClient();
            var url = $"https://api.nuget.org/v3-flatcontainer/{packageId}/{version}/{packageId}.{version}.nupkg";

            var response = await httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();

            var extractor = new PackageAPIExtractor();

            using var responseStream = await response.Content.ReadAsStreamAsync();

            var packageDetails = await extractor.ExtractFromStream(responseStream);

            Approvals.VerifyJson(JsonSerializer.Serialize(packageDetails));
        }

        [Test]
        public async Task ApprovePackageWithNativeLibs()
        {
            var packageId = "boost_contract-vc110";
            var version = "1.71.0";

            var httpClient = new HttpClient();
            var url = $"https://api.nuget.org/v3-flatcontainer/{packageId}/{version}/{packageId}.{version}.nupkg";

            var response = await httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();

            var extractor = new PackageAPIExtractor();

            using var responseStream = await response.Content.ReadAsStreamAsync();

            var packageDetails = await extractor.ExtractFromStream(responseStream);

            Approvals.VerifyJson(JsonSerializer.Serialize(packageDetails));
        }
    }
}
