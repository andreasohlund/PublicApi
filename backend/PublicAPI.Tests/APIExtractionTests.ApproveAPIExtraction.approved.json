{
    "Name": "ExampleAssembly",
    "PublicTypes": [
        {
            "Name": "ISomeInterface",
            "Namespace": "ExampleAssembly",
            "Methods": [],
            "Properties": [],
            "Fields": [],
            "IsInterface": true,
            "Implements": [],
            "BaseType": null,
            "IsStatic": false,
            "IsEnum": false,
            "IsClass": false,
            "IsAbstract": true,
            "IsSealed": false,
            "IsValueType": false
        },
        {
            "Name": "ISomeOtherInterface",
            "Namespace": "ExampleAssembly",
            "Methods": [],
            "Properties": [],
            "Fields": [],
            "IsInterface": true,
            "Implements": [],
            "BaseType": null,
            "IsStatic": false,
            "IsEnum": false,
            "IsClass": false,
            "IsAbstract": true,
            "IsSealed": false,
            "IsValueType": false
        },
        {
            "Name": "PublicEnum",
            "Namespace": "ExampleAssembly",
            "Methods": [],
            "Properties": [],
            "Fields": [
                {
                    "Name": "Value1",
                    "Type": "ExampleAssembly.PublicEnum",
                    "IsStatic": true
                },
                {
                    "Name": "Value2",
                    "Type": "ExampleAssembly.PublicEnum",
                    "IsStatic": true
                },
                {
                    "Name": "value__",
                    "Type": "System.Int32",
                    "IsStatic": false
                }
            ],
            "IsInterface": false,
            "Implements": [],
            "BaseType": "System.Enum",
            "IsStatic": false,
            "IsEnum": true,
            "IsClass": true,
            "IsAbstract": false,
            "IsSealed": true,
            "IsValueType": true
        },
        {
            "Name": "PublicEnumWithExplicitValues",
            "Namespace": "ExampleAssembly",
            "Methods": [],
            "Properties": [],
            "Fields": [
                {
                    "Name": "Value1",
                    "Type": "ExampleAssembly.PublicEnumWithExplicitValues",
                    "IsStatic": true
                },
                {
                    "Name": "Value2",
                    "Type": "ExampleAssembly.PublicEnumWithExplicitValues",
                    "IsStatic": true
                },
                {
                    "Name": "value__",
                    "Type": "System.Int32",
                    "IsStatic": false
                }
            ],
            "IsInterface": false,
            "Implements": [],
            "BaseType": "System.Enum",
            "IsStatic": false,
            "IsEnum": true,
            "IsClass": true,
            "IsAbstract": false,
            "IsSealed": true,
            "IsValueType": true
        },
        {
            "Name": "PublicStruct",
            "Namespace": "ExampleAssembly",
            "Methods": [],
            "Properties": [],
            "Fields": [],
            "IsInterface": false,
            "Implements": [],
            "BaseType": "System.ValueType",
            "IsStatic": false,
            "IsEnum": false,
            "IsClass": true,
            "IsAbstract": false,
            "IsSealed": true,
            "IsValueType": true
        },
        {
            "Name": "SomeAbstractClass",
            "Namespace": "ExampleAssembly",
            "Methods": [],
            "Properties": [],
            "Fields": [],
            "IsInterface": false,
            "Implements": [],
            "BaseType": "System.Object",
            "IsStatic": false,
            "IsEnum": false,
            "IsClass": true,
            "IsAbstract": true,
            "IsSealed": false,
            "IsValueType": false
        },
        {
            "Name": "SomeBaseClass",
            "Namespace": "ExampleAssembly",
            "Methods": [
                {
                    "Name": ".ctor",
                    "ReturnType": "System.Void",
                    "Parameters": [],
                    "IsStatic": false
                }
            ],
            "Properties": [],
            "Fields": [],
            "IsInterface": false,
            "Implements": [],
            "BaseType": "System.Object",
            "IsStatic": false,
            "IsEnum": false,
            "IsClass": true,
            "IsAbstract": false,
            "IsSealed": false,
            "IsValueType": false
        },
        {
            "Name": "SomePublicAttribute",
            "Namespace": "ExampleAssembly",
            "Methods": [
                {
                    "Name": ".ctor",
                    "ReturnType": "System.Void",
                    "Parameters": [],
                    "IsStatic": false
                }
            ],
            "Properties": [],
            "Fields": [],
            "IsInterface": false,
            "Implements": [],
            "BaseType": "System.Attribute",
            "IsStatic": false,
            "IsEnum": false,
            "IsClass": true,
            "IsAbstract": false,
            "IsSealed": false,
            "IsValueType": false
        },
        {
            "Name": "SomePublicClass",
            "Namespace": "ExampleAssembly",
            "Methods": [
                {
                    "Name": ".ctor",
                    "ReturnType": "System.Void",
                    "Parameters": [],
                    "IsStatic": false
                },
                {
                    "Name": "A_ShouldSortPublicMethodsByName",
                    "ReturnType": "System.Void",
                    "Parameters": [],
                    "IsStatic": false
                },
                {
                    "Name": "PublicMethod",
                    "ReturnType": "System.Void",
                    "Parameters": [],
                    "IsStatic": false
                },
                {
                    "Name": "PublicMethodWithOptionalParam",
                    "ReturnType": "System.Void",
                    "Parameters": [
                        {
                            "Name": "optionalParam",
                            "Type": "System.Int32",
                            "IsOptional": true,
                            "Modifier": "none"
                        }
                    ],
                    "IsStatic": false
                },
                {
                    "Name": "PublicMethodWithOutParam",
                    "ReturnType": "System.Void",
                    "Parameters": [
                        {
                            "Name": "outParam",
                            "Type": "System.Object\u0026",
                            "IsOptional": false,
                            "Modifier": "out"
                        }
                    ],
                    "IsStatic": false
                },
                {
                    "Name": "PublicMethodWithParam",
                    "ReturnType": "System.Void",
                    "Parameters": [
                        {
                            "Name": "param",
                            "Type": "System.Object",
                            "IsOptional": false,
                            "Modifier": "none"
                        }
                    ],
                    "IsStatic": false
                },
                {
                    "Name": "PublicMethodWithRefParam",
                    "ReturnType": "System.Void",
                    "Parameters": [
                        {
                            "Name": "refParam",
                            "Type": "System.Object\u0026",
                            "IsOptional": false,
                            "Modifier": "ref"
                        }
                    ],
                    "IsStatic": false
                },
                {
                    "Name": "StaticPublicMethod",
                    "ReturnType": "System.Void",
                    "Parameters": [],
                    "IsStatic": true
                }
            ],
            "Properties": [
                {
                    "Name": "GetSetProperty",
                    "Type": "System.Int32",
                    "HasGetter": true,
                    "HasSetter": true,
                    "IsStatic": false
                },
                {
                    "Name": "StaticProperty",
                    "Type": "System.Int32",
                    "HasGetter": true,
                    "HasSetter": true,
                    "IsStatic": true
                },
                {
                    "Name": "GetOnlyProperty",
                    "Type": "System.Int32",
                    "HasGetter": true,
                    "HasSetter": false,
                    "IsStatic": false
                },
                {
                    "Name": "SetOnlyProperty",
                    "Type": "System.Int32",
                    "HasGetter": false,
                    "HasSetter": true,
                    "IsStatic": false
                }
            ],
            "Fields": [
                {
                    "Name": "PublicField",
                    "Type": "System.Int32",
                    "IsStatic": false
                },
                {
                    "Name": "StaticField",
                    "Type": "System.Int32",
                    "IsStatic": true
                }
            ],
            "IsInterface": false,
            "Implements": [
                "ExampleAssembly.ISomeInterface",
                "ExampleAssembly.ISomeOtherInterface"
            ],
            "BaseType": "ExampleAssembly.SomeBaseClass",
            "IsStatic": false,
            "IsEnum": false,
            "IsClass": true,
            "IsAbstract": false,
            "IsSealed": false,
            "IsValueType": false
        },
        {
            "Name": "SomeStaticClass",
            "Namespace": "ExampleAssembly",
            "Methods": [],
            "Properties": [],
            "Fields": [],
            "IsInterface": false,
            "Implements": [],
            "BaseType": "System.Object",
            "IsStatic": true,
            "IsEnum": false,
            "IsClass": true,
            "IsAbstract": true,
            "IsSealed": true,
            "IsValueType": false
        }
    ],
    "IsNative": false,
    "Version": "1.0.0.0",
    "PublicKey": ""
}